APITaxi/commands/import_zupc.py

```diff
@@ -1,314 +1,388 @@
+    db.session.flush()
+
+    return parent
+
+
+def fill_zupc_special_name(filename, parent_zupc):
+    """File "special_name_insee" from the ZUPC repository can be used to override the name and INSEE code of a ZUPC.
+    """
+    with open(filename) as handle:
+        lines = list(csv.reader(handle))
+
+    if len(lines) != 1:
+        raise ValueError('CSV file %s should contain exactly one line' % filename)
+    insee, name = lines[0]
+
+    parent_zupc.nom = name
+    parent_zupc.insee = insee
+    db.session.add(parent_zupc)
+    db.session.flush()
+
+
+def fill_zupc_update_shape(filename, parent_zupc, action):
+    """Update shape of `parent_zupc` with the shapes defined in `filename`. `action` is a string that is either
+    'include' or 'exclude'.
+
+    If 'include', shapes of `filename` are added to `parent_zupc.shape`.
+    If 'exclude', shapes of `filename` are removed from`parent_zupc.shape`.
+    """
+    assert action in ('include', 'exclude')
+
+    with open(filename) as handle:
+        data = json.load(handle)
+
+    if data['type'] not in ('Feature', 'FeatureCollection'):
+        raise ValueError('Unable to handle geojson type %s in %s' % (data['type'], filename))
+
+    all_shapes = []
+    if data['type'] == 'Feature':  # only one shape
+        all_shapes = [shape(data['geometry'])]
+    else:  # FeatureCollection: list of shapes
+        all_shapes = [shape(part['geometry']) for part in data['features']]
+
+    # new_shape = covering all polygons of all_shapes
+    new_shape = cascaded_union(all_shapes)
+
+    if action == 'include':
+        update = new_shape.union(to_shape(parent_zupc.shape))
+    else:
+        update = new_shape.difference(to_shape(parent_zupc.shape))
+
+    parent_zupc.shape = wkt.dumps(MultiPolygon([update]))
+
+    # wkt.dumps returns a string, refreshing the session fetches the object from database to render parent_zupc.shape as
+    # a geoalchemy2.elements.WKBElement.
+    db.session.refresh(parent_zupc)
+
+    db.session.add(parent_zupc)
+    db.session.flush()
+
+
+def fill_zupc(zupc_dir):
+    """Insert informations into zupc_temp table.
+    """
+    logger.debug('Fill ZUPC temporary table with %s', zupc_dir)
+
+    insee_list_filename = os.path.join(zupc_dir, 'insee.list')
+    if not os.path.exists(insee_list_filename):
+        logger.debug('No insee.list file in %s, nothing to do', zupc_dir)
         return
 
-    z = create_temp_table()
-    load_zupc_temp_table(shape_filename, z)
-    load_zupc(zupc_dir, z)
-    if not confirm_zones():
+    parent_zupc = fill_zupc_union(insee_list_filename)
+    if not parent_zupc:
         return
-    merge_zones(z)
+
+    special_name_filename = os.path.join(zupc_dir, 'special_name_insee')
+    if os.path.exists(special_name_filename):
+        fill_zupc_special_name(special_name_filename, parent_zupc)
+
+    for filename in glob.glob(os.path.join(zupc_dir, '*.include')):
+        fill_zupc_update_shape(filename, parent_zupc, 'include')
```

Comment:
The thing is, what would we do if there is an error?

I think it is better to have a program raising an explicit error if there is a problem rather than silently ignoring it or displaying an error message which could make think that it is a normal behavior.